### DISCUSSION TITLE
Why PIECE_TYPE_NB = 8 ?

### DISCUSSION THREAD

**Discussion by vincentbab** [official-stockfish/Stockfish#5051] [General]
Hello,

I came across this code in `types.h` : 
```c++
enum PieceType {
  NO_PIECE_TYPE, PAWN, KNIGHT, BISHOP, ROOK, QUEEN, KING,
  ALL_PIECES = 0,
  PIECE_TYPE_NB = 8
};
```

I wonder why `PIECE_TYPE_NB = 8` while `7` would be enough ? This could be a tiny improvement in performance (but maybe too small for elo increase). Nonetheless it seems better to not waste memory :)

What do you guys think ?

**Comment by mstembera** [comment]
It's used to initialize some 2D array sizes which are more efficient to index when they are powers of two.

**Reply by vincentbab** [reply]
Thanks for your answer. I found this stackoverflow thread which talk about this: https://stackoverflow.com/questions/9515482/performance-advantages-of-powers-of-2-sized-data

The answer is very interesting and says there might be some down side to power of 2 array size but anyway this seems negligible for stockfish.

But at least I learned something new today, thanks :)
